# Top of /r/femalefashionadvice WAYWT

A script to get all the top comments in the WAYWT of a given month and format them for a post. This was formerly done manually by /u/thecurvynerd (thank you!). Example post: https://www.reddit.com/r/femalefashionadvice/comments/61241t/top_of_waywt_february_2017/

This script does NOT create the imgur album linked to in the title.

The example post, for reference. Jupyter notebook does not render `^` as superscripts.

## [Top of WAYWT: FEBRUARY 2017](http://imgur.com/a/t0468)

> 1\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8oszc/) ^(by */u/Wasmand*) ^**(+162)**  
2\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2j8k6/) ^(by */u/Bloopity_bloop_bloop*) ^**(+141)**  
3\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/dds52c9/) ^(by */u/sohungy*) ^**(+130)**  
4\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd7aw1/) ^(by */u/smolherring*) ^**(+121)**  
5\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd7dgp6/) ^(by */u/shercocked*) ^**(+119)**  
6\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8q9sw/) ^(by */u/sohungy*) ^**(+109)**  
7\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddniypv/) ^(by */u/BlueSpiritBlues*) ^**(+104)**  
8\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd72crh/) ^(by */u/ayiyi*) ^**(+103)**  
9\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhoahf/) ^(by */u/collapsedgovernment*) ^**(+101)**  
10\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddy47x4/) ^(by */u/collapsedgovernment*) ^**(+100)**  
11\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/ddships/) ^(by */u/shercocked*) ^**(+99)**  
12\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddnksu0/) ^(by */u/somethings_nothings*) ^**(+95)**  
13\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2np2z/) ^(by */u/ayiyi*) ^**(+92)**  
14\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8quxg/) ^(by */u/SusannaCarmichael*) ^**(+92)**  
15\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddnkz7v/) ^(by */u/HAPiper*) ^**(+91)**  
16\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd488f/) ^(by */u/BlueSpiritBlues*) ^**(+89)**  
17\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhp9gw/) ^(by */u/MonocleLewinsky*) ^**(+86)**  
18\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/ddsdf5o/) ^(by */u/saphydoodle*) ^**(+86)**  
19\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhq1oh/) ^(by */u/sohungy*) ^**(+83)**  
20\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhyjzl/) ^(by */u/skyllanyx*) ^**(+82)**  
21\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddxw5ba/) ^(by */u/ceruleanmuse*) ^**(+80)**  
22\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd6yr0e/) ^(by */u/villonious*) ^**(+79)**  
23\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/dds54vw/) ^(by */u/BlueSpiritBlues*) ^**(+77)**  
24\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8mjjn/) ^(by */u/muzzyhoo92*) ^**(+75)**  
25\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd73qqj/) ^(by */u/earth_rises*) ^**(+72)**  
26\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/ddshqzt/) ^(by */u/imogenimpulse*) ^**(+72)**  
27\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddxwku5/) ^(by */u/startupwoes*) ^**(+72)**  
28\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de93lv0/) ^(by */u/BlueSpiritBlues*) ^**(+71)**  
29\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8o2fb/) ^(by */u/villonious*) ^**(+69)**  
30\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddy0dxj/) ^(by */u/Scrap-Icon*) ^**(+68)**  
31\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddnrafw/) ^(by */u/villonious*) ^**(+67)**  
32\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2o3h5/) ^(by */u/BlueSpiritBlues*) ^**(+67)**  
33\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd2ivh/) ^(by */u/aftertheswimmingpool*) ^**(+65)**  
34\. [Post](https://www.reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2mlpz/) ^(by */u/averagefruit*) ^**(+65)**  


## Usage

### Knowledge Prerequisites

It will be a lot less painful to use this if you have experience using jupyter notebook. This section expects you know how to download this `ipynb` file (or the whole repository), install dependencies (manually, using a virtualenv, or using conda env) and set environment variables.

### Dependencies

This script requires the following to be installed on your computer:

- [python](https://www.python.org/downloads/) 3.5 (probably any python 3.x would work, but I used 3.5)
- [jupyter notebook](http://jupyter.org/install.html)
- [PRAW](https://praw.readthedocs.io/en/latest/index.html) (the Python Reddit API Wrapper).

There is a `environment.yml` file if you want to create a conda env (`conda env create -f environment.yml`) with jupyter and python 3.5. Conda does not have `praw` as of August 2017, so you will need to `pip install` that manually.

### Environment Variables

The reddit API requires you to register your app to be able to use it. You can get an app ID and app secret by following the steps [here](https://github.com/reddit/reddit/wiki/OAuth2-Quick-Start-Example#first-steps). Select 'script' when it asks what type of app it will be.

Save the app ID, app secret, and your reddit username as [environment variables](https://en.wikipedia.org/wiki/Environment_variable#Assignment). How you set these depends on what OS you're using.

```
export REDDIT_CLIENT_ID=your_id
export REDDIT_CLIENT_SECRET=your_secret
export REDDIT_USER=your_username
```

### Run the script

In the cell that says

```
# read-only subreddit instance
ffa = reddit.subreddit('femalefashionadvice')
# 1-indexed months. e.g. 2 would be February
MONTH = 2
YEAR = 2017
# How many comments you want in the 'Top of' list.
LIMIT = 34
```

Set `MONTH`, `YEAR`, and `LIMIT` to the values you want.

Unfortunately this script still requires manual intervention to filter out WAYWT posts within the month time range that are not, in fact, WAYWT posts. You can select the cell *under* the one that says `print_titles(month_submissions_with_waywt)`, and then from the notebook toolbar select `Cell` > `Run all above` to run all the cells up to that point. It should print a list of the submissions that contain 'WAYWT' in the title and were posted in the month you set. Here you can see that you would want to filter out things like previous "top of" posts. 

You can then change the variable `not_waywt = [3, 4, 7, 8, 10]` to have the numbers of the posts that you *don't* want to include. With that cell selected, you can then do `Run all below` to run the rest of the script.

You can paste the output of the `print(top_of_waywt_post)` cell in the bottom markdown cell to see what it would look like with formatting applied.

Note that deleted comments (if a comment author is `None`) are not included.

The code in this notebook is not the cleanest or most optimized because I am lazy, but hopefully it is still useful.

In [1]:
import calendar
import os

from datetime import datetime

import praw

In [2]:
def get_all_submissions_for_month(month, year):
    """
    Params
    ------
    month: int
    year: int
    """
    # https://stackoverflow.com/questions/42950/get-last-day-of-the-month-in-python
    last_day_of_month = calendar.monthrange(YEAR, MONTH)[1]

    unix_start_month = datetime.timestamp(datetime(YEAR, MONTH, 1))
    unix_end_month = datetime.timestamp(datetime(YEAR, MONTH, last_day_of_month))
    
    return [post for post in ffa.submissions(start=unix_start_month, end=unix_end_month, extra_query=None)]


In [3]:
def is_waywt(submission):
    return ('WAYWT' in submission.title) and ('Announcement' not in submission.title)

In [4]:
def print_titles(submissions):
    # print for human checking
    if not submissions:
        print("No submissions to print!")
    else:
        for i,submission in enumerate(submissions):
            print('{0}. {1}'.format(i, submission.title))

In [5]:
def extract_comments(submission):
    # This will exclude the 'more comments' objects
    submission.comments.replace_more(limit=0)
    # Manually sort because setting submission.comment_sort to 'top' does not seem to actually sort by upvotes
    return [top_level_comment for top_level_comment in submission.comments if top_level_comment.author is not None]

In [6]:
def sort_comments_by_upvotes(comments):
    """comments is a list of pram Comments"""
    return sorted(comments, key=lambda x: x.ups, reverse=True)

In [7]:
def get_all_comments(submissions):
    all_comments = []
    for submission in submissions:
        all_comments += extract_comments(submission)
    return all_comments

In [8]:
def print_top_n_comments(limit, comments):
    """
    Params
    ------
    limit: int
    """
    info = "Author: /u/{0}\nUpvotes: {1}\nPermalink: {2}\n\n"
    for comment in comments[:limit]:
        print(info.format(comment.author, comment.ups, 'https://reddit.com'+comment.permalink()))

In [9]:
# Instantiate your read-only reddit instance using PRAW
# You should have these set as environment variables
reddit = praw.Reddit(client_id=os.environ.get('REDDIT_CLIENT_ID'),
                     client_secret=os.environ.get('REDDIT_CLIENT_SECRET'),
                     user_agent=os.environ.get('REDDIT_USER'))

In [10]:
# read-only subreddit instance
ffa = reddit.subreddit('femalefashionadvice')
# 1-indexed months. e.g. 2 would be February
MONTH = 2
YEAR = 2017
# How many comments you want in the 'Top of' list.
LIMIT = 34

In [11]:
month_submissions = get_all_submissions_for_month(MONTH, YEAR)

In [12]:
month_submissions_with_waywt = [submission for submission in month_submissions if is_waywt(submission)]

In [13]:
print_titles(month_submissions_with_waywt)

0. WAYWT - February 26, 2017
1. WAYWT - February 22, 2017
2. WAYWT - February 19, 2017
3. /r/TeenFA's Top of WAYWT: January 2017
4. Top of WAYWT: January 2017
5. WAYWT - February 15, 2017
6. WAYWT - February 12, 2017
7. Obscuring Your Face in WAYWT Pics (x-post r/malefashionadvice)
8. Month Long WAYWT Album Thread!
9. WAYWT - February 08, 2017
10. REMINDER: Month Long WAYWT ends this Wednesday 2/8! Thread at 12PM EST!(x-post MFA)
11. WAYWT - February 05, 2017
12. WAYWT - February 01, 2017


In [14]:
# You probably need to manually intervene here
# Note that you need to stick to the zero-indexing as printed above.
not_waywt = [3, 4, 7, 8, 10]
month_waywts = []
for i,submission in enumerate(month_submissions_with_waywt):
    if i not in not_waywt:
        month_waywts.append(submission)

print_titles(month_waywts)

0. WAYWT - February 26, 2017
1. WAYWT - February 22, 2017
2. WAYWT - February 19, 2017
3. WAYWT - February 15, 2017
4. WAYWT - February 12, 2017
5. WAYWT - February 08, 2017
6. WAYWT - February 05, 2017
7. WAYWT - February 01, 2017


In [15]:
# this might take a couple of seconds
all_top_level_comments = get_all_comments(month_waywts)

In [16]:
sorted_comments = sort_comments_by_upvotes(all_top_level_comments)

In [17]:
print_top_n_comments(LIMIT, sorted_comments)

Author: /u/Wasmand
Upvotes: 178
Permalink: https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8oszc


Author: /u/sohungy
Upvotes: 141
Permalink: https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/dds52c9


Author: /u/smolherring
Upvotes: 128
Permalink: https://reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd7aw1


Author: /u/shercocked
Upvotes: 118
Permalink: https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd7dgp6


Author: /u/sohungy
Upvotes: 108
Permalink: https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8q9sw


Author: /u/ayiyi
Upvotes: 105
Permalink: https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd72crh


Author: /u/BlueSpiritBlues
Upvotes: 103
Permalink: https://reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddniypv


Author: /u/SusannaCarmichael
Upvotes: 97
Permalink: http

In [18]:
def create_post_text(comments):
    title = "## TOP OF WAYWT: {0} {1}\n".format(calendar.month_name[MONTH], YEAR)

    # See https://www.reddit.com/wiki/commenting
    table_header = "Rank | Comment | User | Upvotes\n" + \
                   "---- | ---- | ---- | -------\n"

    row = "{rank} | [{post_title}](https://reddit.com{permalink}) | /u/{user} | +{upvotes}\n"

    post = title + '\n' + table_header

    # You could speed this up by refactoring to keep the submission title
    # with each comment, or by just making the permalink text just say 'link'.
    for i, comment in enumerate(comments):
        post += row.format(
            rank=i+1,
            permalink=comment.permalink(),
            user=comment.author,
            upvotes=comment.ups,
            post_title=reddit.submission(id=comment._extract_submission_id()).title
        )
    return post

In [19]:
# This will take a little while
top_of_waywt_post = create_post_text(sorted_comments[:LIMIT])

In [20]:
print(top_of_waywt_post)

## TOP OF WAYWT: February 2017

Rank | Comment | User | Upvotes
---- | ---- | ---- | -------
1 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8oszc) | /u/Wasmand | +178
2 | [WAYWT - February 15, 2017](https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/dds52c9) | /u/sohungy | +141
3 | [WAYWT - February 05, 2017](https://reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd7aw1) | /u/smolherring | +128
4 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd7dgp6) | /u/shercocked | +118
5 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8q9sw) | /u/sohungy | +108
6 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd72crh) | /u/ayiyi | +105
7 | [WAYWT - February 12, 2017](https://reddit.com/r/fem

Below is the output of the above in markdown, for human-checking that it came out right.

The script-generated output does not match with the original post. I think it's because the original post omitted the February 12th WAYWT. The upvotes do not match due to a time difference in when the data was collected (at the time people could have gone and upvoted the comments after seeing them via the 'Top of' post). /u/Bloopity_bloop_bloop deleted their comment since the original 'Top of' post, and it did not get included here.

## TOP OF WAYWT: February 2017

Rank | Comment | User | Upvotes
---- | ---- | ---- | -------
1 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8oszc) | /u/Wasmand | +177
2 | [WAYWT - February 15, 2017](https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/dds52c9) | /u/sohungy | +140
3 | [WAYWT - February 05, 2017](https://reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd7aw1) | /u/smolherring | +128
4 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd7dgp6) | /u/shercocked | +119
5 | [WAYWT - February 12, 2017](https://reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddniypv) | /u/BlueSpiritBlues | +108
6 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8q9sw) | /u/sohungy | +106
7 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd72crh) | /u/ayiyi | +102
8 | [WAYWT - February 15, 2017](https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/ddships) | /u/shercocked | +97
9 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8quxg) | /u/SusannaCarmichael | +94
10 | [WAYWT - February 22, 2017](https://reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2np2z) | /u/ayiyi | +94
11 | [WAYWT - February 12, 2017](https://reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddnksu0) | /u/somethings_nothings | +91
12 | [WAYWT - February 15, 2017](https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/ddsdf5o) | /u/saphydoodle | +87
13 | [WAYWT - February 05, 2017](https://reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd488f) | /u/BlueSpiritBlues | +86
14 | [WAYWT - February 08, 2017](https://reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhp9gw) | /u/MonocleLewinsky | +85
15 | [WAYWT - February 08, 2017](https://reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhq1oh) | /u/sohungy | +82
16 | [WAYWT - February 19, 2017](https://reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddxw5ba) | /u/ceruleanmuse | +79
17 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd6yr0e) | /u/villonious | +77
18 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8mjjn) | /u/muzzyhoo92 | +76
19 | [WAYWT - February 15, 2017](https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/dds54vw) | /u/BlueSpiritBlues | +76
20 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de93lv0) | /u/BlueSpiritBlues | +73
21 | [WAYWT - February 15, 2017](https://reddit.com/r/femalefashionadvice/comments/5u89ok/waywt_february_15_2017/ddshqzt) | /u/imogenimpulse | +73
22 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd73qqj) | /u/earth_rises | +71
23 | [WAYWT - February 19, 2017](https://reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddy0dxj) | /u/Scrap-Icon | +68
24 | [WAYWT - February 22, 2017](https://reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2o3h5) | /u/BlueSpiritBlues | +67
25 | [WAYWT - February 12, 2017](https://reddit.com/r/femalefashionadvice/comments/5tm3q4/waywt_february_12_2017/ddnrafw) | /u/villonious | +67
26 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8o2fb) | /u/villonious | +66
27 | [WAYWT - February 22, 2017](https://reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2mlpz) | /u/averagefruit | +65
28 | [WAYWT - February 05, 2017](https://reddit.com/r/femalefashionadvice/comments/5s847m/waywt_february_05_2017/ddd2ivh) | /u/aftertheswimmingpool | +64
29 | [WAYWT - February 08, 2017](https://reddit.com/r/femalefashionadvice/comments/5stiil/waywt_february_08_2017/ddhoq9q) | /u/earth_rises | +63
30 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd70bkl) | /u/useafuckingcoaster | +63
31 | [WAYWT - February 22, 2017](https://reddit.com/r/femalefashionadvice/comments/5vjhx0/waywt_february_22_2017/de2j361) | /u/omiiii | +61
32 | [WAYWT - February 19, 2017](https://reddit.com/r/femalefashionadvice/comments/5uyvqj/waywt_february_19_2017/ddxz0s8) | /u/BlueSpiritBlues | +59
33 | [WAYWT - February 01, 2017](https://reddit.com/r/femalefashionadvice/comments/5rgaam/waywt_february_01_2017/dd6zgdz) | /u/-lyrota- | +59
34 | [WAYWT - February 26, 2017](https://reddit.com/r/femalefashionadvice/comments/5waqux/waywt_february_26_2017/de8pwc3) | /u/somethings_nothings | +58